### PR TITLE
Add support for arrays of fundamental types

### PR DESCRIPTION
**Pull request by sdroege**
https://github.com/gtk-rs/gir/issues/388

### CODE REVIEW

**Comment by sdroege** [comment]
https://github.com/sdroege/glib-rs/commit/2e85eba71c826fc589e53df021f21a0b9ba1a362 for an example how this looks like. g_filename_from_utf8() should probably be ignored.

**Comment by EPashkin** [comment]
`g_key_file_set_double_list` and `g_key_file_set_integer_list` as it don't gir don't have `*` in arrays.
So generated ffi contains wrong `list`
`pub fn g_key_file_set_double_list(key_file: *mut GKeyFile, group_name: *const c_char, key: *const c_char, list: c_double, length: size_t);
`

**Comment by sdroege** [comment]
Yes, see my GLib commit that disables these two. The .gir file is wrong

**Comment by sdroege** [comment]
Also this currently does not seem to (always?) work with out parameters (g_file_load_contents() e.g.).

**Comment by EPashkin** [comment]
`filename_from_utf8` compiles normally under windows but IMHO really don't needed.

**Comment by EPashkin** [comment]
Other compile problems only starts with gtk, so all previous builds fine.

**Comment by sdroege** [comment]
The support for out arrays needs various functions everywhere to be disabled that have missing/wrong annotations.

**Comment by EPashkin** [comment]
From manual gtk\list_store maybe better change `values: &[&glib::Value]` to `values: &[&ToValue]` for input arrays ?

**Comment by sdroege** [comment]
That can be added at a later time. That it is generated now is just a side effect of allowing out-arrays :)

**Comment by sdroege** [comment]
Most remaining compile errors seem to be the following:
1) Wrong annotations
2) "zero-terminated" arrays of fundamental types, we can't really handle them
3) arrays of fundamental types without a length anywhere (some also compile but they would be unsafe as-is)
4) gdk::Atom curiousities

I'm checking now how to handle 2) and 3)

**Comment by EPashkin** [comment]
Hm, now we stumbled upon reserved names names

```rust
pub fn base64_decode_step(in_: &[u8], state: &mut i32, save: &mut u32) -> (usize, Vec<u8>) {
    let len = in.len() as usize;
    unsafe {
        let mut out = ptr::null_mut();
        let ret = ffi::g_base64_decode_step(in_.to_glib_none().0, len, &mut out, state, save);
        (ret, FromGlibPtrContainer::from_glib_full(out))
    }
}
```

**Comment by sdroege** [comment]
That one is not generated anymore

**Comment by EPashkin** [comment]
@sdroege, please add `let array_name = nameutil::mangle_keywords(&array_name[..]).into_owned();` in  body of `if let Some(array_name) = array_name {` in `function_parameters.rs` even if you ignore this function

**Comment by sdroege** [comment]
Sure

**Comment by EPashkin** [comment]
Sorry `.into_owned()` unneeded

File: src/analysis/rust_type.rs [hunk 1]
```diff
@@ -147,6 +147,40 @@ fn rust_type_full(
                     format!("Vec<{}>", s)
                 })
         }
+        CArray(inner_tid) if ConversionType::of(&env.library, inner_tid) == ConversionType::Direct => {
```
Comment (EPashkin) [inline src/analysis/rust_type.rs]: Strange, this case under `CArray(inner_tid)` without `if`. It dont skipped?

Comment (sdroege) [inline src/analysis/rust_type.rs]: It's also with if. Only pointers are handled above

Comment (EPashkin) [inline src/analysis/rust_type.rs]: You right, sorry.

**Comment by EPashkin** [comment]
@sdroege PR ready for merge?

**Comment by sdroege** [comment]
Yes, good to go IMHO. All remaining failures are bugs in the .gir and can be ignored (will do so when doing the regenerate PRs).

**Comment by EPashkin** [comment]
Thanks, @sdroege

**Pull request merged**